File: lambda_function/function/lambda.js
const AWS = require('aws-sdk');
const s3 = new AWS.S3();

exports.handler = async (event) => {
    const sourceBucket = 'cca-pbl'; // Replace with your source bucket name
    const destinationBucket = 'backup-bucket'; // Replace with your destination bucket name
    
    try {
        const objects = await s3.listObjectsV2({ Bucket: sourceBucket }).promise();

        for (const obj of objects.Contents) {
            // Delete the existing object in the destination bucket, if it exists
            try {
                await s3.deleteObject({ Bucket: destinationBucket, Key: obj.Key }).promise();
            } catch (err) {
                // Ignore errors if the object doesn't exist in the destination bucket
            }

            // Copy the object from the source bucket to the destination bucket
            const copyParams = {
                CopySource: `${sourceBucket}/${obj.Key}`,
                Bucket: destinationBucket,
                Key: obj.Key,
            };

            await s3.copyObject(copyParams).promise();
        }
        
        const response = {
            statusCode: 200,
            body: JSON.stringify('S3 bucket backup completed successfully.'),
        };
        
        return response;
    } catch (error) {
        console.error('Error:', error);
        throw error;
    }
};

File: python/lambda.py
import boto3
import json
from datetime import datetime

def lambda_handler(event, context):
    source_bucket = "cca-pbl"
    backup_bucket = "cca-pbl-backup"
    
    s3 = boto3.client('s3')

    # Retrieve the list of objects in the main bucket
    response = s3.list_objects_v2(Bucket=source_bucket)

    # Backup each object to the backup bucket
    for obj in response.get('Contents', []):
        key = obj['Key']

        # Copy the object to the backup bucket
        s3.copy_object(
            Bucket=backup_bucket,
            CopySource={'Bucket': source_bucket, 'Key': key},
            Key=key
        )

    # Create a log file with a timestamp
    timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    log_content = f"Backup completed at {timestamp}"
    
    s3.put_object(
        Bucket=backup_bucket,
        Key=f"backup_logs/{timestamp}_log.txt",
        Body=log_content
    )

    return {
        'statusCode': 200,
        'body': json.dumps('Backup successful!')
    }
